yb-voyager/cmd/config.go

```diff
@@ -0,0 +1,467 @@
+	 2. Loads the config file if explicitly provided via --config, or defaults to ~/.yb-voyager.yaml.
+	 3. Reads and validates the configuration file for allowed global keys, sections, and section keys.
+	 4. Binds Viper config values to Cobra flags, giving priority to command-line flags over config values.
+	 5. Returns an error if config file reading, validation, or flag binding fails.
+
+	This setup ensures CLI > ENV > Config precedence and allows environment variables
+	prefixed with YB_VOYAGER_ to override config values.
+*/
+func initConfig(cmd *cobra.Command) error {
+	v := viper.New()
+
+	if cfgFile != "" {
+		// Use config file from the flag.
+		v.SetConfigFile(cfgFile)
+	} else {
+		// Find home directory.
+		home, err := os.UserHomeDir()
+		cobra.CheckErr(err)
+
+		// Search config in home directory with name ".yb-voyager" (without extension).
+		v.AddConfigPath(home)
+		v.SetConfigType("yaml")
+		v.SetConfigName(".yb-voyager")
+	}
+
+	// If a config file is found, read it in.
+	if err := v.ReadInConfig(); err == nil {
+		fmt.Println("Using config file:", v.ConfigFileUsed())
+	} else {
+		if _, ok := err.(viper.ConfigFileNotFoundError); !ok {
+			return err
+		}
+	}
+
+	// Validate the config file for allowed keys and sections
+	err := validateConfigFile(v)
+	if err != nil {
+		return err
+	}
+
+	// Bind the config values to the Cobra command flags
+	err = bindCobraFlagsToViper(cmd, v)
+	if err != nil {
+		return fmt.Errorf("failed to bind cobra flags to viper: %w", err)
+	}
+
+	return nil
+}
+
+/*
+validateConfigFile checks the loaded configuration for correctness.
+
+	It performs the following validations:
+	1. Ensures that all global-level keys (non-nested) are in the allowed list.
+	2. Ensures that all section names (e.g., export-schema) are known and valid.
+	3. Ensures that all nested keys inside each section are valid for that section.
+	4. Ensures that only one section from each group of mutually exclusive aliases is used (e.g.,
+	"export-data" vs "export-data-from-source").
+
+	Any invalid global keys, unknown sections, or invalid section keys are collected and printed.
+	Conflicts between mutually exclusive sections are also reported.
+	If any validation error is found, an error is returned.
+
+	This helps catch misconfigurations early and provides comprehensive feedback to the user.
+*/
+func validateConfigFile(v *viper.Viper) error {
+	var (
+		invalidGlobalKeys   []string
+		invalidSectionKeys  = make(map[string][]string)
+		invalidSections     []string
+		conflictingSections = [][]string{}
+		presentSections     = make(map[string]bool)
+	)
+
+	fmt.Println("Validating config file...")
+	for _, key := range v.AllKeys() {
+		parts := strings.Split(key, ".")
+		if len(parts) == 1 {
+			// Check global level keys
+			if !allowedGlobalConfigKeys[key] {
+				invalidGlobalKeys = append(invalidGlobalKeys, key)
+			}
+		} else {
+			// Validate section-based keys
+			section, nestedKey := parts[0], parts[1]
```

Comment:
Maybe not immediately pressing to incorporate, just a note for future: 

Today, we have only one level of nesting so this is okay. We store `allowedConfigSections` and validate the sections that are allowed, and the keys under that as well. 
But what if we have something like:
```
a: 
   b: 
      c: 
```

How do we validate `c` ? 

A more flexible way would be to just store all allowed viper keys in a single list: 
"a.b", "a.c", "a.d", "b.c", "b.c.d", etc. If you can't find the key, just throw an error saying unknown configuration parameter, and point the user to a sample config file.

Reply:
that makes sense!